keps/sig-auth/3325-self-user-attributes-review-api/README.md

```diff
@@ -0,0 +1,357 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+There is no resource which represents a user in Kubernetes. Instead, Kubernetes has authenticators to get user attributes from tokens or x509 certificates or by using the OIDC provider or receiving them from the external webhook. This KEP proposes adding a new API endpoint to see what attributes the current user has after the authentication.
+
+## Motivation
+
+Authentication is complicated, especially made by [proxy](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy) or [webhook](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) authenticators or their combinations.
+It may be obscure which user attributes the user eventually gets after all that magic happened before authentication.
+The motivation for this KEP is to reduce obscurity and help users with debugging the authentication stack.
+
+### Goals
+
+- Add the API endpoint to get user attributes
+
+### Non-Goals
+
+- Add a corresponding kubectl command
+
+## Proposal
+
+Add a new API endpoint to the `authentication` group - `SelfUserAttributesReview`.
+The user will hip the endpoint after authentication happens, so all attributes will be available to return.
+
+## Design Details
+
+This design is inspired by the `*AccessReview` and `TokenReview` APIs.
+The endpoint has no input parameters or a `spec` field because only the authentication result is required.
+
+### Request
+
+The structure for building a request:
+```go
+type SelfUserAttributesReview struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Status is filled in by the server with the user attributes.
+	Status SelfUserAttributesReviewStatus `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
```

Comment:
Will this ever have a status field? If not, I'd skip this wrapper.

Reply:
I used the status here due to two reasons:
1. To be in line with other *Review APIs. A user sends a request with an empty status and receives the same object with the filled status fields.
2. To be able to easily extend this API in the future. I thought we would like to add some `spec`s, e.g., to request only some attributes:

   ```yaml
   apiVersion: authentication.k8s.io/v1alpha1
   kind: SelfSubjectAttributesReviewStatus
   spec:
     returnOnly:
     - name
     - groups
   ```
   In this case, having a wrapper gives us some advantages.

Hope I get your question right.

Reply:
I prefer to keep the status field primarily because it feels consistent with the existing APIs.